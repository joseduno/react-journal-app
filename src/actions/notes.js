import Swal from 'sweetalert2'
import {db} from '../firebase/config'
import { fileUpload } from '../helpers/fileUpload'
import { loadNotes } from '../helpers/loadNotes'
import { types } from '../types/types'


export const startNewNote = () => {
    /* getState es una funcion similar al useSelector */ 
    return async(dispatch, getState) => {
        try {
            const {uid} = getState().auth
            const newNote = {
                title: '',
                body: '',
                date: new Date().getTime(),
            }
            const doc = await db.collection(`${uid}/journal/notes`).add(newNote)
            dispatch(activeNote(doc.id, newNote))
            dispatch(addNewNote(doc.id, newNote))
            
        } catch (error) {
            console.log(error)
        }
    }
}


export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})


export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {id, ...note}
})


export const startLoadingNotes = uid => {
    return async(dispatch) => {
        try {
            const notes = await loadNotes(uid)
            dispatch(setNotes(notes))
            
        } catch (error) {
            console.log(error)
        }
    }
}


export const setNotes = notes => ({
    type: types.notesLoad,
    payload: notes
})


export const startSavingNote = note => {
    return async(dispatch, getState) => {
        try {
            const {uid} = getState().auth
            const noteToFirestore = {...note}
            delete noteToFirestore.id
            if (!noteToFirestore.url) {
                delete noteToFirestore.url
            }
            await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore)
            dispatch(refreshNote(note))
            /* Comentado porque en entorno de pruebas me salta un error
            No se hacer el mock del Swal */
            Swal.fire({
                icon: 'success',
                title: 'Saved!',
                text: note.title
            })
            
        } catch (error) {
            console.log(error)
        }
    }
}


export const refreshNote = note => ({
    type: types.notesUpdated,
    payload: note
})


export const startUploadingFile = file => {
    return async(dispatch, getState) => {
        try {
            const {active:activeNote} = getState().notes
            /* Comentado porque en entorno de pruebas me salta un error
            No se hacer el mock del Swal */
            Swal.fire({
                title: 'Uploading...',
                text: 'Please wait...',
                willOpen: () => {
                    Swal.showLoading()
                }
            })
            const fileUrl = await fileUpload(file)
            dispatch(startSavingNote({...activeNote, url: fileUrl}))
            /* Comentado porque en entorno de pruebas me salta un error
            No se hacer el mock del Swal */
            Swal.close()
            
        } catch (error) {
            console.log(error)
        }
    }
}


export const startDeletingNote = id => {
    return async(dispatch, getState) => {
        try {
            const {uid} = getState().auth
            await db.doc(`${uid}/journal/notes/${id}`).delete()
            dispatch(noteDelete(id))
            
        } catch (error) {
            console.log(error)
        }
    }
}


export const noteDelete = id => ({
    type: types.notesDelete,
    payload: id
})


export const logoutCleaning = () => ({
    type: types.notesLogoutCleaning
})