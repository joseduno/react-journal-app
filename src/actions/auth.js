import Swal from 'sweetalert2'
import { firebase, googleAuthProvider } from "../firebase/config"
import { types } from "../types/types"
import { logoutCleaning } from './notes'
import { finishLoading, startLoading } from "./ui"



/* inicio de sesion con firebase */
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading())
        // No entiendo este return, se usa para que las pruebas funciones, pero no entiendo por que
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({user}) => {
                dispatch(login(user.uid, user.displayName))
                dispatch(finishLoading())
            })
            .catch(({message}) => {
                Swal.fire('', message, 'error')
                dispatch(finishLoading())
            })
    }
}


/* inicio de sesion con google */
export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                const {uid, displayName} = user
                dispatch(login(uid, displayName))
            })
            .catch(({message}) => {
                Swal.fire('', message, 'error')
            })
    }
}

/* registro con firebase */
export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {

        // para registrar
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async({user}) => {

                // para actualizar datos
                await user.updateProfile({
                    displayName: name
                })
                dispatch(login(user.uid, user.displayName))
            })
            .catch(({message}) => {
                Swal.fire('', message, 'error')
            })
    }
}

export const startLogout = () => {
    return async(dispatch) => {
        // para cerrar sesion
        try {
            await firebase.auth().signOut()
            dispatch(logout())
            dispatch(logoutCleaning())
        } catch ({message}) {
            Swal.fire('', message, 'error')
        }
    }
}

/* Estas son funciones que retornan acciones!, las anteriores son funciones para 
procesar informacion y posteriormente con un dispatch, ejecutar funciones que retornan 
una accion */
export const login = (uid, displayName) => ({
    type: types.login, 
    payload: {
        uid,
        displayName
    }
})

export const logout = () => ({
    type: types.logout
})