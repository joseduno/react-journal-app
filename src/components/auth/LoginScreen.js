import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import useForm from '../../hooks/useForm'


export const LoginScreen = () => {

    const {loading} = useSelector(state => state.ui)

    /* Hook de react-redux para disparar acciones y que el 
    'store' se encargue de buscar el reducer adecuado */
    const dispatch = useDispatch()

    const [formValues, handleChange] = useForm({
        email: '',
        password: ''
    })
    const {email, password} = formValues

    const handleSubmit = e => {
        e.preventDefault();
        dispatch(startLoginEmailPassword(email, password))
    }

    /* el dispatch del custom hook de react-redux puede recibir funciones como 
    arguemntos que retornen un callback, el hook pasara la funcion 'dispatch' 
    como argumento a ese callback retornado, para que desde ese callback, pueda ser llamando 
    finalmente la accion (con su type y payload), ver 'actions/auth.js' */
    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin())
    }

    return (
        <div className='animate__animated animate__fadeIn animate__faster'>
            <h3 className='auth__title'>Login</h3>
            <form onSubmit={handleSubmit}>
                <input
                    type='text'
                    placeholder='Email'
                    name='email'
                    className='auth__input'
                    autoComplete='off'
                    value={email}
                    onChange={handleChange}
                />
                <input
                    type='password'
                    placeholder='Password'
                    name='password'
                    className='auth__input'
                    value={password}
                    onChange={handleChange}
                />
                <button
                    type='submit'
                    className='btn btn-primary btn-block'
                    disabled={loading}
                >
                    Login
                </button>

                <div className='auth__social-network'>
                    <div 
                        className="google-btn"
                        onClick={handleGoogleLogin}
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link to='/auth/register' className='link'>
                    Create new account
                </Link>
            </form>
        </div>
    )
}
