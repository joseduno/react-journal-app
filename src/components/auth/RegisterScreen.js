import React from 'react'
import { Link } from 'react-router-dom'
import useForm from '../../hooks/useForm'
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';


export const RegisterScreen = () => {

    const dispatch = useDispatch()

    /* useSelector: hook del react-redux utilizado para acceder a la 
    informacion del store (tal como se ve en la pestana state del 
    redux devtools) */
    const {msgError} = useSelector(state => state.ui)

    const [formValues, handleChange] = useForm({
        name: '',
        email: '',
        password: '',
        password2: '',
    })

    const {name, email, password, password2} = formValues

    const handleSubmit = e => {
        e.preventDefault()
        if (isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email, password, name))
        }
    }

    const isFormValid = () => {
        if (name.trim().length === 0) {
            dispatch(setError('Name is required'))
            return false
        } else if (!validator.isEmail(email)) {
            dispatch(setError('Email is not valid'))
            return false
        } else if (password !== password2 || !validator.isStrongPassword(password)) {
            dispatch(setError('Password is not valid'))
            return false
        }
        dispatch(removeError())
        return true
    }

    

    return (
        <div className='animate__animated animate__fadeIn animate__faster'>
            <h3 className='auth__title'>Register</h3>
            <form onSubmit={handleSubmit}>
                {
                    msgError
                    &&
                    (<div className='auth__alert-error'>
                        {msgError}
                    </div>)
                }
                <input
                    type='text'
                    placeholder='Name'
                    name='name'
                    className='auth__input'
                    autoComplete='off'
                    value={name}
                    onChange={handleChange}
                />
                <input
                    type='text'
                    placeholder='Email'
                    name='email'
                    className='auth__input'
                    autoComplete='off'
                    value={email}
                    onChange={handleChange}
                />
                <input
                    type='password'
                    placeholder='Password'
                    name='password'
                    className='auth__input'
                    value={password}
                    onChange={handleChange}
                />
                <input
                    type='password'
                    placeholder='Confirm password'
                    name='password2'
                    className='auth__input'
                    value={password2}
                    onChange={handleChange}
                />
                <button
                    type='submit'
                    className='btn btn-primary btn-block mb-5'
                >
                    Sign in
                </button>

                <Link to='/auth/login' className='link'>
                    Already registered?
                </Link>

            </form>
        </div>
    )
}
