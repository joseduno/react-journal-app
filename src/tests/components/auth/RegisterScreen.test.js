import { mount } from "enzyme"
import { Provider } from "react-redux"
import { MemoryRouter } from "react-router-dom"
import configureStore from "redux-mock-store"
import thunk from "redux-thunk"
import { startRegisterWithEmailPasswordName } from "../../../actions/auth"
import { RegisterScreen } from "../../../components/auth/RegisterScreen"


jest.mock('../../../actions/auth', () => ({
    startRegisterWithEmailPasswordName: jest.fn()
}))


const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initialState = {
    ui: {
        msgError: null
    }
}
let store = mockStore(initialState)
store.dispatch = jest.fn()

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
)


describe('Pruebas en componente <RegisterScreen />', () => {

    beforeEach(() => {
        store = mockStore(initialState)
        jest.clearAllMocks()
    })

    test('Debe hacer match con el snapshot ', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('Debe iniciar el proceso de registro con email password y name', () => {
        wrapper.find('input[name="name"]').simulate('change', {target: {
            name: 'name',
            value: 'test'
        }})
        wrapper.find('input[name="email"]').simulate('change', {target: {
            name: 'email',
            value: 'test@test.com'
        }})
        wrapper.find('input[name="password"]').simulate('change', {target: {
            name: 'password',
            value: 'Test123-,.'
        }})
        wrapper.find('input[name="password2"]').simulate('change', {target: {
            name: 'password2',
            value: 'Test123-,.'
        }})
        wrapper.find('form').simulate('submit', {preventDefault(){}})
        expect(startRegisterWithEmailPasswordName).toHaveBeenCalledWith(
            'test@test.com', 'Test123-,.', 'test'
        )
    })

    test('Debe mostrar alerta de error', () => {
        const initialState = {
            ui: {
                loading: false,
                msgError: 'Email no es correcto'
            }
        }
        const store = mockStore(initialState)
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        )
        const divError = wrapper.find('.auth__alert-error')
        expect(divError.exists()).toBe(true)
        expect(divError.text().trim()).toBe('Email no es correcto')
    })
    
    
    
})
