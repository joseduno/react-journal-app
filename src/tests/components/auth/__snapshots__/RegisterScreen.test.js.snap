// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente <RegisterScreen /> Debe hacer match con el snapshot  1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster"
>
  <h3
    className="auth__title"
  >
    Register
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Sign in
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>
</div>
`;
